File: apps/web/src/api/lib/ulid.ts
/**
 * Minimal ULID generator for Cloudflare Workers.
 *
 * ULIDs are 26-char, Crockford base32-encoded, time-sortable unique IDs.
 * Format: 10 chars timestamp (48-bit ms) + 16 chars randomness (80-bit).
 */

const ENCODING = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"; // Crockford base32

function encodeTime(now: number, len: number): string {
  let str = "";
  for (let i = len; i > 0; i--) {
    const mod = now % 32;
    str = ENCODING[mod] + str;
    now = (now - mod) / 32;
  }
  return str;
}

function encodeRandom(len: number): string {
  const bytes = crypto.getRandomValues(new Uint8Array(len));
  let str = "";
  for (let i = 0; i < len; i++) {
    str += ENCODING[bytes[i] % 32];
  }
  return str;
}

/**
 * Generate a new ULID. Uses `crypto.getRandomValues` (available in Workers).
 */
export function ulid(): string {
  const now = Date.now();
  return encodeTime(now, 10) + encodeRandom(16);
}

File: packages/sdk-go/ulid.go
package statsfactory

import (
	"crypto/rand"
	"time"
)

// Crockford base32 encoding alphabet.
const ulidEncoding = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"

// generateULID creates a ULID (26-char, time-sortable unique ID).
// Uses crypto/rand for the random component.
func generateULID() string {
	now := time.Now().UnixMilli()

	// Encode 48-bit timestamp as 10 Crockford base32 chars.
	var buf [26]byte
	for i := 9; i >= 0; i-- {
		buf[i] = ulidEncoding[now%32]
		now /= 32
	}

	// Encode 80 bits of randomness as 16 Crockford base32 chars.
	var rb [16]byte
	_, _ = rand.Read(rb[:])
	for i := 0; i < 16; i++ {
		buf[10+i] = ulidEncoding[rb[i]%32]
	}

	return string(buf[:])
}
